Infrastructure for known functions, types and constructors

jschneider-bensch opened the pull request:
This PR adds more infrastructure to recognize known types, constructors and constructor patterns.
It also provides placeholder stubs for inserting pre-defined translations of these in a follow up.

Fixes #449.

W95Psp commented on engine/names/src/lib.rs:
```diff
@@ -1,11 +1,43 @@
+
+    let _ = hash(HashAlgorithm::Sha256, &bytes);
+    let _ = cipher_text.clone();
+    // End crypto abstractions example
```
You can put this code in a separate `crypto_abstractions.rs` (and add a `mod crypto_abstractions;` to that file) if you like, it would be nicer than begin-end comments IMO :)

And you could `use hax_lib_protocol::crypto::*` only there.

W95Psp commented on engine/backends/proverif/proverif_backend.ml:
```diff
@@ -115,35 +115,207 @@ module Print = struct
     [
-      (Core__ops__try_trait__Try__branch, fun args -> empty);
-      (* just an example *)
+      (* Core dependencies *)
```
What do you plan on doing with those?
Do you plan to rename each of those manually to something else?

Since you are building the library on ProVerif's side as well, can't you just name functions with a scheme we can directly produce, without the indirection of rewriting all those symbols?

W95Psp replied:
Well, I mean that instead of manually saying `hax_lib_protocol::cal::hmac` is actually `foobar` in the proverif side, why not calling that `cal_hmac` in proverif, and just translate all the item `hax_lib_protocol::cal::foo` as `cal_foo`? or even, why not use a fully qualified name there? i.e. `hax_lib_protocol__cal__hmac`

jschneider-bensch replied:
Ah, I think I understand now, thanks! I'm wondering if that would only work if we just want translate one for one to matching names and not to something more complicated on the ProVerif side.
An example may be when we want to drop some of the arguments of a function when translating to ProVerif, then we would need special logic for that translation in particular. Sorry, this is a bit abstract right now. Do you think I could address it properly when I work on #447 and #448, where I have to fill in the translations?

W95Psp commented on engine/backends/proverif/proverif_backend.ml:
```diff
@@ -115,35 +115,207 @@ module Print = struct
 
   let translate_known_function fname args =
-    (List.find_exn ~f:(assoc_known_function fname) library_functions |> snd)
+    (List.find_exn ~f:(assoc_known_name fname) library_functions |> snd) args
+
+  let translate_known_constructor cname fields =
+    (List.find_exn ~f:(assoc_known_name cname) library_constructors |> snd)
+      fields
+
+  let translate_known_constructor_pat cname args =
+    (List.find_exn ~f:(assoc_known_name cname) library_constructors_patterns
+    |> snd)
       args
 
+  let translate_known_type tname =
+    List.find_exn ~f:(assoc_known_name tname) library_types |> snd
+
   let is_known_function fname =
-    List.exists ~f:(assoc_known_function fname) library_functions
+    List.exists ~f:(assoc_known_name fname) library_functions
+
+  let is_known_constructor cname =
+    List.exists ~f:(assoc_known_name cname) library_constructors
+
+  let is_known_type tname =
+    List.exists ~f:(assoc_known_name tname) library_types
```
I know those `translate_known_*` functions are called only after calling their associated `is_known_*` functions, but I don't like to much those `find_exn`...
Maybe instead of doing `match _ with | ... where is_known_* -> code`, we should `match _ with | ... -> (match List.find ... with | Some ... -> code | _ -> None)`?

jschneider-bensch replied:
Thanks for this suggestion, I'll try that.
Perhaps I can also refactor this more anyway, since it's all basically the same logic.

jschneider-bensch replied:
I've simplified this a bit and removed these `find_exn`s.

W95Psp replied:
Cool, thanks!

jschneider-bensch commented:
I realized I had introduced a regression on translating `Result` constructor by messing up the order of match arms. I fixed it by changing the order now, but should revisit when I work on #448, so it uses the same translation mechanism as the other `core` names.
